### DISCUSSION TITLE
Filter or block open source snippets by license type

### DISCUSSION THREAD

**Discussion by TomPADonaldson** [community/community#51118] [Copilot Conversations]
### Select Topic Area

General

### Body

I am trying out Copilot in a JetBrains environment. Part of the setup process asked if I would like to allow or block code snippets from open source projects.

But all open source licenses are not equal. In particular I want to avoid snippets from code that is licensed under any GNU license.  For now I have blocked all code snippets.

How can I be more selective, that is, permit/block code snippets based on their license without having to manually search for the license for each snippet?

If license based filtering of snippets is not currently a feature of Copilot, will it be included in future releases? I think this is something that may greatly impact the usefulness of Copilot, or at least to the degree that snippets are useful.

Thanks,
Tom

**Comment by Akash1134** [comment]
Hi @TomPADonaldson ,

Thank you for posting your query in the Copilot community. 

Currently, the functionality to filter snippets based on the license is not available. Please note that GitHub Copilot is a generative AI, meaning that it doesn't copy and paste open-source code, so there is no list of code snippets for it to check against to find the license but this is something we are aware of.

**Reply by TomPADonaldson** [reply]
Perhaps I am confused by the setup question regarding allowing or blocking snippets from open source. 

You you saying that it is not directly copying from open source, but rather generating code snippets based on its own derivation of a code pattern from the snippets?

If so, that puts us one step beyond the most blatant IP infraction. But maybe not far enough to get past a forensic software analysis. For example, changing variable names does not hide the software pattern/algorithm from a skilled examiner (human), and is probably even less effective in hiding the source of a snippet or larger piece of code from a human driving an AI pattern detector.

Glad it is being considered! But none of us want to be part of the precedent-setting multi-year multi-million dollar litigation that could result from a miscalculation.

Love what Copilot does. Keep up the good work! And help us compute safely.
